### PR TITLE
DAOS-9311 build: build MPI-related test binaries with docker

### PR DESCRIPTION
**Pull request by knard-intel**
Dockerfile RUN use /bin/sh to run command. As indicated in the bash manual: "A non-interactive shell invoked with the name sh does not attempt to read any other startup files."
More information could be found at https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html

Building MPI-related test binaries of DAOS (e.g. daos_perf, daos_test, etc.) needs some environment variables releated to modules tool.  These environmant variables are  defined in /etc/profile.d/modules.sh.

Do not hesitate to add reviewers if I forgot someone: I used git blame to create the list of reviewers.

### CODE REVIEW

**Comment by ashleypittman** [comment]
I don't understand this PR, firstly it's against 2.0 rather than master but also mpi is picked up from the mpicc binary being on path and does appear to be happening.

There is a "--require-optional" option to scon-3 that we've added, in order to force optional components to be included, if there's concern that we're not picking up mpi then we should add this and resolve any issues that fall out of it.  It does look like I've checked only Ubuntu and the ticket references CentOS 8.

**Comment by ashleypittman** [comment]
See also https://github.com/daos-stack/daos/blob/master/site_scons/env_modules.py#L42-L49 for our MODULEPATH handing, if the ticket is right then one solution would be to add the path to MODULEPATH here, as we do for other paths.

**Comment by ashleypittman** [comment]
In fact this patch also resolves the issue.  We'd probably want to duplicate rather than replace the existing value.

[20:21] [255] (master)* ampittma@ethenaff-MOBL:~/coral/daos $ git diff
diff --git a/site_scons/env_modules.py b/site_scons/env_modules.py
index d9b933a01..44603d52d 100644
--- a/site_scons/env_modules.py
+++ b/site_scons/env_modules.py
@@ -41,7 +41,7 @@ class _env_module(): # pylint: disable=invalid-name
         # in it's MODULEPATH, for some unknown reason
         os.environ["MODULEPATH"] = ":".join([os.path.sep +
                                              os.path.join("usr", "share",
-                                                          "modules"),
+                                                          "modulefiles"),
                                              os.path.sep +
                                              os.path.join("etc",
                                                           "modulefiles")] +

**Comment by knard-intel** [comment]
> I don't understand this PR, firstly it's against 2.0 rather than master but also mpi is picked up from the mpicc binary being on path and does appear to be happening.
> 
> There is a "--require-optional" option to scon-3 that we've added, in order to force optional components to be included, if there's concern that we're not picking up mpi then we should add this and resolve any issues that fall out of it. It does look like I've checked only Ubuntu and the ticket references CentOS 8.

 I am not aware of all the details, but from what I understood, Kevan Rehm (from HPE) noticed that MPI-related test binaries of DAOS (e.g. daos_perf, daos_test, etc.) where not compiled any more with the Dockerfiles `utils/docker/Dockerfile.xxxx`. After more investigation, he discovered that it was a problem with the `MODULEPATH` environment variables which was not defined when the RUN statement of a Dockerfile was used.  He also find a workaround with adding the following statement in the Dockerfile: `ENV MODULEPATH /usr/share/modulefiles`.  He also noticed, that this issue was not happening when the compilation was done with a bash interractive shell.

From my side, I was able to reproduce this issue and find that the modules associated tools was expecting to have some environment files defined thanks to the profile files `/etc/profile.d/modules.[sh,csh] `. The issue with the Dockerfile `RUN` statement is that it is using a `/bin/sh` non interractive shell (i.e. seems to be hardcoded in Docker and could not be changed) which is not loading any any startup files encompassing the ones located in `/etc/profile.d`. To overcome this issue,  my idea was to load the default modules configuration file `/etc/profile.d/modules.sh`. By this way, all the environment variables needed by modules utils tools should be well defined. In a first time, I targeted the `release/2.0` branch as it was indicated as one of the _Fix version_ in the ticket.

From my side, I am not against adding the required path `/usr/share/modulefiles` in the `site_scons/env_modules.py` as you proposed. However, I think that loading the module `/etc/profile.d/modules.[sh,csh] ` (when it exists) before running `scons` command is a good practice in order to have an environment equivalent as if we are running `scons` in an interractive shell. Moreover, with this solution we will automatically manage if some new paths or environment variables are needed by future versions of modules tools.

Hope, this explanation will help.

**Comment by ashleypittman** [comment]
https://github.com/daos-stack/daos/commit/eccd94849c38b730c524100ddcdb80f3c0448e92 seems to be where the origional logic in env_modules.py came from although it doesn't add much.

I'm wary of having similar MODULEPATH handling in two completely different locations in the source tree, it makes it very difficult to debug and work on the source when that happens, the fix for this issue would appear to be the same as we applied for leap before.  Docker isn't the only place where building is done automatically so it's not clear how/where Jenkins or the RPM build would be picking up the default environment from either.

Having /etc/profiles/modules.sh setup for scons commands in the dockerfiles does have advantages as you say, but there's also a danger that it gets set for some scons commands and not others, leading to inconsistencies about what gets built at the different states however, so I think the ENV MODULEFILE option would be safer although as you'll note doesn't handle the case where the contents of the profile script change.

Ultimately I think this code probably changes infrequently enough that the env_modules change is enough, --require-optional would be a good thing in the dockerfiles as well.  At the time being for release/2.0 we're landing stuff to master, then backporting as required.

I've incorported the env_modules.py change into https://github.com/daos-stack/daos/pull/7677 and picked up your other bug fixes as well, but it's still building so I haven't been able to verify the results yet.

**Comment by knard-intel** [comment]
> [eccd948](https://github.com/daos-stack/daos/commit/eccd94849c38b730c524100ddcdb80f3c0448e92) seems to be where the origional logic in env_modules.py came from although it doesn't add much.
> 
> I'm wary of having similar MODULEPATH handling in two completely different locations in the source tree, it makes it very difficult to debug and work on the source when that happens, the fix for this issue would appear to be the same as we applied for leap before. Docker isn't the only place where building is done automatically so it's not clear how/where Jenkins or the RPM build would be picking up the default environment from either.
> 
> Having /etc/profiles/modules.sh setup for scons commands in the dockerfiles does have advantages as you say, but there's also a danger that it gets set for some scons commands and not others, leading to inconsistencies about what gets built at the different states however, so I think the ENV MODULEFILE option would be safer although as you'll note doesn't handle the case where the contents of the profile script change.
> 
> Ultimately I think this code probably changes infrequently enough that the env_modules change is enough, --require-optional would be a good thing in the dockerfiles as well. At the time being for release/2.0 we're landing stuff to master, then backporting as required.
> 
> I've incorported the env_modules.py change into #7677 and picked up your other bug fixes as well, but it's still building so I haven't been able to verify the results yet.

I agree that they should not be so frequent modification of the default modules tools environment variables and having the same work done in two separate places is not a good thing for maintenance.  Thus, what I propose is to check that your patch is fixing this issue. If yes, I will cancel this PR.  Is it fine for you ?

**Comment by knard-intel** [comment]
@ashleypittman , I just finished to test your proposal and it works well :)
Thus, I am going to cancel this PR in favor of your PR https://github.com/daos-stack/daos/pull/7677

**Comment by knard-intel** [comment]
Reopen pull request with using solution https://github.com/daos-stack/daos/pull/7677 (https://daosio.atlassian.net/browse/DAOS-7821)

File: site_scons/env_modules.py [hunk 1]
```diff
@@ -39,14 +39,10 @@ def __init__(self):
...
         # in it's MODULEPATH, for some unknown reason
-        os.environ["MODULEPATH"] = ":".join([os.path.sep +
-                                             os.path.join("usr", "share",
-                                                          "modules"),
-                                             os.path.sep +
-                                             os.path.join("etc",
-                                                          "modulefiles")] +
-                                            os.environ.get("MODULEPATH",
-                                                           "").split(":"))
+        os.environ["MODULEPATH"] = ":".join([os.path.join("/usr", "share", "modules"),
```
Comment (ashleypittman) [inline site_scons/env_modules.py:42]: This version worked but I had also updated it since to use os.sep

Comment (knard-intel) [reply site_scons/env_modules.py:42]: Arghhh, I did not check my cherry pick :'(

Comment (knard-intel) [reply site_scons/env_modules.py:42]: Going to fix it asap

Comment (knard-intel) [reply site_scons/env_modules.py]: Fixed and checked with your branch.
My apologies.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Are you planning to close this?   Your comment implies as much so just getting it out of my list.  Please re-add me if that's a mistake.

**Comment by knard-intel** [comment]
@ashleypittman do you have any other remarks regarding this PR ?
@johannlombardi, @tanabarr, @JohnMalmberg, @jolivier23, could you help me to review this PR.
I would like to land this Pull request asap as it is tagged for the 2.0 release.

**Comment by knard-intel** [comment]
> Are you planning to close this? Your comment implies as much so just getting it out of my list. Please re-add me if that's a mistake.

Re-added you, not yet familiar with gitlab and did not realize that closing and reopening pull request have closed review requests.  Thanks to have let me realize that.

**Review by ashleypittman — APPROVED** [review]
Fine with me.  This is onto the 2.0 branch but the changes have been tested on master in https://github.com/daos-stack/daos/pull/7677 which is also in review.

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**